bliss/models/location_encoder.py

```diff
@@ -352,28 +358,32 @@ def _encode_for_n_sources(
             `n_samples x n_ptiles x max_detections x ...`
         """
         tile_n_sources = tile_n_sources.clamp(max=self.max_detections)
-        n_ptiles = var_params_flat.size(0)
-        vpf_padded = F.pad(var_params_flat, (0, self.dim_out_all, 0, 0))
 
+        # first, we transform `tile_n_sources` so that it can be used as an index
+        # for looking up detections in `params_per_source`
+        ts2 = rearrange(tile_n_sources, "ns b nth ntw -> ns (b nth ntw)")
+        ts3 = self.n_detections_map[ts2]
+        ts4 = rearrange(ts3, "ns np md -> np (ns md) 1")
+        ts5 = ts4.expand(ts4.size(0), ts4.size(1), self.n_params_per_source)
+
+        # next, we pad `params_per_source` with a dummy column of zeros that will be looked up
+        # (copied) whenever fewer the `max_detections` sources are present. `gather` does the copy.
+        pps2 = rearrange(params_per_source, "b nth ntw td pps -> (b nth ntw) td pps")
+        pps3 = F.pad(pps2, (0, 0, 0, 1))
+        pps4 = torch.gather(pps3, 1, ts5)
+        pps5 = rearrange(pps4, "np (ns md) pps -> ns np md pps", ns=tile_n_sources.size(0))
+
+        # finally, we slice pps5 by parameter group because these groups are treated differently,
```

Comment:
Could we just use `torch.split()` for this part? E.g.
```
dims = param.values()
var_params_split = torch.split(pp5, dims, -1)
var_params_for_n_sources = {k: v for k,v in zip(dims.keys(), var_params_split)}
```

Reply:
I like it!